tests/unit/library/test_json.py

```diff
@@ -2023,7 +2041,180 @@ def test_ignored_keys_allowed_as_properties():
+        target_obj = {"a": "hello", "b": 42, "c": True, **extra_items}
+        generate_and_check(target_obj, schema_obj)
+
+    @pytest.mark.parametrize(
+        "bad_obj",
+        [
+            # Missing one
+            ({"a": "hello", "b": 42}),
+            ({"a": "hello", "c": True}),
+            ({"b": 42, "c": True}),
+            # Missing two
+            ({"a": "hello"}),
+            ({"b": 42}),
+            ({"c": True}),
+            # Missing all
+            ({}),
+        ]
+    )
+    def test_all_required_bad(self, bad_obj):
+        schema_obj = {**self.schema_obj, "required": self.ALL_REQUIRED}
+        check_match_failure(
+            bad_string=_to_compact_json(bad_obj),
+            schema_obj=schema_obj,
+            compact=True
+        )
+
+    @pytest.mark.parametrize(
+        "extra_items",
+        [
+            {},
+            {"d": "hello"},
+            {"d": 42, "e": True},
+        ]
+    )
+    @pytest.mark.parametrize(
+        "required",
+        SOME_REQUIRED_SUBSETS,
+    )
+    def test_some_required_good(self, required, extra_items):
+        base_obj = {"a": "hello", "b": 42, "c": True}
+        schema_obj = {**self.schema_obj, "required": required}
+
+        for subset in self.SOME_REQUIRED_SUBSETS:
+            if set(required).issubset(subset):
+                target_obj = {**{key: base_obj[key] for key in subset}, **extra_items}
+                generate_and_check(target_obj, schema_obj)
+
+    @pytest.mark.parametrize(
+        "required",
+        SOME_REQUIRED_SUBSETS,
+    )
+    def test_some_required_bad(self, required):
+        base_obj = {"a": "hello", "b": 42, "c": True}
+        schema_obj = {**self.schema_obj, "required": required}
+
+        for subset in self.SOME_REQUIRED_SUBSETS:
+            if not set(required).issubset(subset):
+                bad_obj = {key: base_obj[key] for key in subset}
+                check_match_failure(
+                    bad_string=_to_compact_json(bad_obj),
+                    schema_obj=schema_obj,
+                    compact=True
+                )
+
+    # No equivalent "bad" tests for none required, as the schema is satisfied by an empty object
+    @pytest.mark.parametrize(
+        "extra_items",
+        [
+            {},
+            {"d": "hello"},
+            {"d": 42, "e": True},
+        ]
+    )
+    @pytest.mark.parametrize(
+        "target_obj",
+        [
+            {},
+            {"a": "hello"},
+            {"b": 42},
+            {"c": True},
+            {"a": "hello", "b": 42},
+            {"a": "hello", "c": True},
+            {"b": 42, "c": True},
+            {"a": "hello", "b": 42, "c": True},
+        ]
+    )
+    def test_none_required(self, target_obj, extra_items):
+        schema_obj = {**self.schema_obj, "required": self.NONE_REQUIRED}
+        generate_and_check({**target_obj, **extra_items}, schema_obj)
+
+class TestRequiredPropertiesScaling:
```

Comment:
The assertions about N*(N-1)/2 and 2N-1 come from some back of the envelope theoretical calculations I made in the above thread. I seem to be off by a small constant factor in each. Tests are framed as `<=` expected just in case someone wants to step in and improve things. These tests may be a bit fragile and overly dependent on private details, but they definitely gave me some peace of mind. Thoughts @riedgar-ms ?